main/src/ca/uwaterloo/flix/language/phase/Uncurrier2.scala

```diff
@@ -0,0 +1,366 @@
+          val e = uncurryApply(exp)
+          Expression.Universal(fparam, e, loc)
+
+        case Expression.Cast(exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.Cast(e, tpe, loc)
+
+        case Expression.TryCatch(exp, rules, tpe, loc) =>
+          val e = uncurryApply(exp)
+          val rs = rules map {
+            case CatchRule(sym, clazz, body) =>
+              val b = uncurryApply(body)
+              CatchRule(sym, clazz, b)
+          }
+          Expression.TryCatch(e, rs, tpe, loc)
+
+        case Expression.InvokeConstructor(constructor, args, tpe, loc) =>
+          val as = args map uncurryApply
+          Expression.InvokeConstructor(constructor, as, tpe, loc)
+
+        case Expression.InvokeMethod(method, exp, args, tpe, loc) =>
+          val e = uncurryApply(exp)
+          val as = args map uncurryApply
+          Expression.InvokeMethod(method, e, as, tpe, loc)
+
+        case Expression.InvokeStaticMethod(method, args, tpe, loc) =>
+          val as = args map uncurryApply
+          Expression.InvokeStaticMethod(method, as, tpe, loc)
+
+        case Expression.GetField(field, exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.GetField(field, e, tpe, loc)
+
+        case Expression.PutField(field, exp1, exp2, tpe, loc) =>
+          val e1 = uncurryApply(exp1)
+          val e2 = uncurryApply(exp2)
+          Expression.PutField(field, e1, e2, tpe, loc)
+
+        case Expression.GetStaticField(field, tpe, loc) => Expression.GetStaticField(field, tpe, loc)
+
+        case Expression.PutStaticField(field, exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.PutStaticField(field, e, tpe, loc)
+
+        case Expression.NewChannel(exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.NewChannel(e, tpe, loc)
+
+        case Expression.GetChannel(exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.GetChannel(e, tpe, loc)
+
+        case Expression.PutChannel(exp1, exp2, tpe, loc) =>
+          val e1 = uncurryApply(exp1)
+          val e2 = uncurryApply(exp2)
+          Expression.PutChannel(e1, e2, tpe, loc)
+
+        case Expression.SelectChannel(rules, default, tpe, loc) =>
+          val rs = rules map {
+            case SelectChannelRule(sym, chan, exp) =>
+              val c = uncurryApply(chan)
+              val e = uncurryApply(exp)
+              SelectChannelRule(sym, c, e)
+          }
+          val d = default.map(uncurryApply)
+          Expression.SelectChannel(rs, d, tpe, loc)
+
+        case Expression.ProcessSpawn(exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.ProcessSpawn(e, tpe, loc)
+
+        case Expression.ProcessPanic(msg, tpe, loc) => Expression.ProcessPanic(msg, tpe, loc)
+
+        case Expression.FixpointConstraintSet(cs0, tpe, loc) => Expression.FixpointConstraintSet(cs0, tpe, loc)
```

Comment:
Must recurse here.

Reply:
I suppose this also applies to renameBoundVars?